File: android/app/src/main/java/com/example/cryptobenchmark/environment/discovery/ServiceKey.java
package com.example.cryptobenchmark.environment.discovery;

import java.util.Locale;

/** Composite because algorithm names contain '/', '.', ':' and '#', so no flat string id is safe. */
public final class ServiceKey {

    private final String type;
    private final String algorithm;

    private ServiceKey(String type, String algorithm) {
        this.type = type;
        this.algorithm = algorithm;
    }

    public static ServiceKey of(String type, String algorithm) {
        return new ServiceKey(fold(type), fold(algorithm));
    }

    /**
     * Locale.ROOT or a Turkish-locale device folds "Cipher" to "CIPHER" with a dotted I and stops
     * matching. On Android the default locale is the user's, so this is not hypothetical.
     */
    private static String fold(String value) {
        return value == null ? null : value.toUpperCase(Locale.ROOT);
    }

    public String getType() {
        return type;
    }

    public String getAlgorithm() {
        return algorithm;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        if (!(other instanceof ServiceKey)) {
            return false;
        }
        ServiceKey that = (ServiceKey) other;
        return (type == null ? that.type == null : type.equals(that.type))
                && (algorithm == null ? that.algorithm == null : algorithm.equals(that.algorithm));
    }

    @Override
    public int hashCode() {
        int result = type == null ? 0 : type.hashCode();
        return 31 * result + (algorithm == null ? 0 : algorithm.hashCode());
    }

    @Override
    public String toString() {
        return type + "." + algorithm;
    }
}

File: environment/discovery/src/main/kotlin/contract/ServiceKey.kt
package io.github.junekim0007.cryptobench.discovery.contract

import java.util.Locale

internal class ServiceKeyException(message: String) : IllegalArgumentException(message)

internal class ServiceKey(type: String, algorithm: String) {

    val type: String = validated("type", type)

    val algorithm: String = validated("algorithm", algorithm)

    override fun equals(other: Any?): Boolean =
        other is ServiceKey && type == other.type && algorithm == other.algorithm

    override fun hashCode(): Int = 31 * type.hashCode() + algorithm.hashCode()

    override fun toString(): String = "$type.$algorithm"

    companion object {

        fun fold(value: String): String = value.uppercase(Locale.ROOT)

        private fun validated(field: String, value: String): String {
            if (value.isBlank()) {
                throw ServiceKeyException("missing_field: $field")
            }
            return fold(value)
        }
    }
}
